src/l500/l500-options.cpp

```diff
@@ -255,75 +396,230 @@ namespace librealsense
+
+    void l500_preset_option::set( float value )
+    {
+        if( static_cast< rs2_l500_visual_preset >( int( value ) )
+            == RS2_L500_VISUAL_PRESET_DEFAULT )
+            throw invalid_value_exception( to_string()
+                                           << "RS2_L500_VISUAL_PRESET_DEFAULT was deprecated! " );
+
+        verify_max_usable_range_restrictions( RS2_OPTION_VISUAL_PRESET, value );
+        _owner->change_preset( static_cast< rs2_l500_visual_preset >( int( value ) ) );
+
+        float_option_with_description< rs2_l500_visual_preset >::set( value );
+    }
+
+    void l500_preset_option::set_value( float value )
+    {
+        float_option_with_description< rs2_l500_visual_preset >::set( value );
+    }
+
+    void l500_preset_option::verify_max_usable_range_restrictions( rs2_option opt, float value ) 
+    {
+        // Block changing visual preset while Max Usable Range is on [RS5-8358]
+        if( _owner->get_depth_sensor().supports_option( RS2_OPTION_ENABLE_MAX_USABLE_RANGE )
+            && ( _owner->get_depth_sensor().get_option( RS2_OPTION_ENABLE_MAX_USABLE_RANGE ).query()
+                 == 1.0 ) )
+        {
+            if( ( RS2_OPTION_VISUAL_PRESET == opt )
+                && ( value == RS2_L500_VISUAL_PRESET_MAX_RANGE ) )
+                return;
+
+            throw wrong_api_call_sequence_exception(
+                to_string()
+                << "Visual Preset cannot be changed while Max Usable Range is enabled" );
+        }
+    }
+
+    void l500_options::on_set_option( rs2_option opt, float value )
+    {
+        auto advanced_controls = get_advanced_controls();
+        if( std::find( advanced_controls.begin(), advanced_controls.end(), opt )
+            != advanced_controls.end() )
         {
-            auto advanced_controls = get_advanced_controls();
-            if (std::find(advanced_controls.begin(), advanced_controls.end(), opt) != advanced_controls.end())
+            if (opt == RS2_OPTION_DIGITAL_GAIN)
             {
-                verify_max_usable_range_restrictions( opt, value );
-                move_to_custom();
+                // WA for fw bug will be removed after fixed on FW
+                //std::this_thread::sleep_for( std::chrono::milliseconds( 500 ) );
+                update_defaults();
             }
-            else
-                throw  wrong_api_call_sequence_exception(to_string() << "on_set_option support advanced controls only "<< opt<<" injected");
+                
+
+            // when we moved to auto preset we set all controls to -1 
+            // so we have to set preset controls to defaults values now
+            /*auto curr_preset = ( rs2_l500_visual_preset )(int)_preset->query();
+            if( curr_preset == RS2_L500_VISUAL_PRESET_AUTOMATIC )
+               set_preset_controls_to_defaults();*/
+
+            move_to_custom();
+
+            if( opt != RS2_OPTION_DIGITAL_GAIN )
+                _hw_options[opt]->set_manually( true );
```

Comment:
Can do this for digital_gain, too

Reply:
?

Reply:
it's relevant only for l500_hw_options to identify if we are on automatic preset